<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Detect Damage in Single Cell RNA Sequencing Data â€¢ DamageDetective</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Detect Damage in Single Cell RNA Sequencing Data">
<meta name="description" content="Simulate datasets containing a specified proportion of damaged cells, estimate the distribution of damage in a dataset and detect the cells in a dataset that are most likely to be damaged given the similarity to simulated damaged cells.">
<meta property="og:description" content="Simulate datasets containing a specified proportion of damaged cells, estimate the distribution of damage in a dataset and detect the cells in a dataset that are most likely to be damaged given the similarity to simulated damaged cells.">
<meta property="og:image" content="https://alicenjoyhenning.github.io/DamageDetective/logo.svg">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">DamageDetective</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/DamageDetective.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.svg" class="logo" alt=""><h1 id="damagedetective-">DamageDetective <a class="anchor" aria-label="anchor" href="#damagedetective-"></a>
</h1>
</div>
<!-- badges: start -->

<!-- badges: end -->
<div class="section level2">
<h2 id="content">Content<a class="anchor" aria-label="anchor" href="#content"></a>
</h2>
<p><a href="#description">Description</a> | <a href="#installation">Installation</a> | <a href="#quick-start">Quick start</a> | <a href="#authors">Authors</a> | <a href="#license">License</a> | <a href="#references">References</a></p>
</div>
<div class="section level2">
<h2 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h2>
<p>Damaged cells are a class of low-quality artifacts in single-cell RNA sequencing (scRNA-seq) describing cells that have succumbed to stress before being sequenced. As this results in altered gene expression profiles, retaining damaged cells in downstream analyses compromises the reliability of the analysis. The detection and removal of damaged cells is therefore a standard practice in the pre-processing of scRNA-seq data.</p>
<p>Current approaches detect damage according to deviations in cell-level quality control metrics influenced by the loss of plasma membrane integrity, a well-accepted consequence of damage. However, this approach assumes all viable cells follow similar distributions, an assumption that does not always hold in heterogeneous samples. More recent methods improve upon this by analysing cells at a population level, isolating cells with similar distributions before assessing finer deviations. However, this assumes that all distinct populations represent true cells meaning if abundant, damaged cells risk misclassification as true cells. Ultimately, the filtering decisions of current approaches are motivated more by statistical definitions of deviation than biological definitions of damage, a non-linear, stochastic process that varies across cell types.</p>
<p><code>DamageDetective</code> takes a different approach inspired by <code>DoubletFinder</code><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mo>*</mo></msup><annotation encoding="application/x-tex">^*</annotation></semantics></math>, a high performing community accepted tool for doublet removal in scRNA-seq. Here, rather than detecting damage by measuring the extent to which cells deviate from each other, it is detected by measuring the extent to which cells align with artificially damaged versions of themselves. Grounded, too, by loss of plasma membrane integrity, this approach simulates damage through the probabilistic escape of cytoplasmic RNA where the proportion of escape is assumed to directly reflect damage severity. By comparing the expression profiles of true cells to artificially damaged cell profiles in reduced-dimensional space, <code>DamageDetective</code> estimates the damage severity of true cells in a score from 0 to 1. This provides an intuitive scale for filtering, based directly on biological definitions of damage, that is standardised across cell types, samples, and experiments.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>DamageDetective</code> can be installed from CRAN using,</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'DamageDetective'</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, the latest development version can be installed directly from GitHub using,</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AlicenJoyHenning/DamageDetective"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>To verify installation, run the following to see if you can view the package vignette and the function help page,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alicenjoyhenning.github.io/DamageDetective/">DamageDetective</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"DamageDetective"</span><span class="op">)</span></span>
<span><span class="op">?</span><span class="fu">DamageDetective</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The demonstrations below can be followed immediately after loading the package and serve as a test to ensure all is running smoothly. For function descriptions and usage examples please refer to the <a href="link">package vignette</a>.</p>
<p>Begin by loading the dummy data provided by the package, <code>test_counts</code>, an artificial PBMC dataset containing 500 cells and 10009 genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"test_counts"</span>, package <span class="op">=</span> <span class="st">"DamageDetective"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">test_counts</span><span class="op">)</span></span>
<span><span class="co"># [1] 10009   500</span></span></code></pre></div>
<ul>
<li>
<h3>
Return default output
</h3>
</li>
</ul>
<p>The primary goal of <code>DamageDetective</code> is to inform the filtering of damaged cells from single cell data. This can be achieved using the <code>detect_damage</code> function that requires a count matrix as input and returns a data frame containing the barcodes of the count matrix with the estimated levels of damage.</p>
<p>The results can be accessed from the <code>output</code> slot and used to inform filtering of cells from the count matrix. Below we are filtering cells with a damage level above 70 %,</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform damage detection</span></span>
<span><span class="va">default_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/detect_damage.html">detect_damage</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">default_test</span><span class="op">$</span><span class="va">qc_summary</span><span class="op">)</span> </span>
<span><span class="co"># </span></span>
<span></span>
<span><span class="co"># Filter cells with an estimated damage level above 70 % </span></span>
<span><span class="va">undamaged_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">default_test</span><span class="op">$</span><span class="va">qc_summary</span>, <span class="va">DamageDetective</span> <span class="op">&gt;</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="va">filtered_counts</span> <span class="op">&lt;-</span> <span class="va">test_counts</span><span class="op">[</span>, <span class="va">undamaged_cells</span><span class="op">]</span></span></code></pre></div>
<blockquote>
<p>By default, <code>detect_damage</code> will provide plots of the data where each cell is a point coloured according to the damage level estimated by <code>DamageDetective</code>. This will be stored in <code>default_test$plot</code> as a <code>ggplot2</code> object that can be manipulated using <code>ggplot2</code> functionality. To disable plotting, indicate using the <code>generate_plot = FALSE</code> argument.</p>
</blockquote>
<ul>
<li>
<h3>
Return filtered output
</h3>
</li>
</ul>
<p>Alternatively, instead of returning an annotated data frame, <code>detect_damage</code> can return the filtered count matrix <code>filter_threshold = TRUE</code>. Here, just as above, filtering is done according to a threshold for the estimated level of damage, specified using the <code>filter_threshold</code> parameter. By default, <code>filter_threshold = 0.75</code>. Essentially, this provides a means of automating the filtering process shown above.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform damage detection &amp; filtering according to non-default threshold</span></span>
<span><span class="va">filtered_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/detect_damage.html">detect_damage</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>, </span>
<span>  filter_counts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filter_threshold <span class="op">=</span> <span class="fl">0.7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">filtered_test</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">filtered_test</span><span class="op">)</span></span>
<span><span class="co">#</span></span></code></pre></div>
<p>To explore these parameters and more, you can visit the package vignette or via the function help page,</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="fu"><a href="reference/detect_damage.html">detect_damage</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p><code>DamageDetective</code> is made available for public use through the <a href="https://opensource.org/licenses/AGPL-3.0" class="external-link">GNU AGPL-3.0</a> license.</p>
</div>
<div class="section level2">
<h2 id="authors">Authors<a class="anchor" aria-label="anchor" href="#authors"></a>
</h2>
<p><strong>Alicen Henning</strong><br>
Stellenbosch University, Cape Town, South Africa<br>
Bioinformatics and Computational Biology</p>
<p>This work was done under the supervision of Prof Marlo MÃ¶ller, Prof Gian van der Spuy, and Prof AndrÃ© Loxton.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>McGinnis, C. S., Murrow, L. M., &amp; Gartner, Z. J. (2019). DoubletFinder: Doublet Detection in Single-Cell RNA Sequencing Data Using Artificial Nearest Neighbors. <em>Cell Systems, 8</em>(4), 329-337.e4. <a href="https://doi.org/10.1016/j.cels.2019.03.003" class="external-link">https://doi.org/10.1016/j.cels.2019.03.003</a>
</li>
</ul>
<hr>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>AGPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing DamageDetective</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Alicen Henning <br><small class="roles"> Author, maintainer, copyright holder </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cosimameyer/overviewR/actions" class="external-link"><img src="https://github.com/AlicenJoyHenning/DamageDetective/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alicen Henning.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
