<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Detecting Damaged Cells in Single-Cell RNA Sequencing Data • DamageDetective</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Detecting Damaged Cells in Single-Cell RNA Sequencing Data">
<meta name="description" content="Detects and filters damaged cells in single-cell RNA sequencing (scRNA-seq) data using a novel approach inspired by DoubletFinder. Damage is detected by measuring the extent to which cells deviate from artificially damaged profiles of themselves, simulated through the probabilistic escape of cytoplasmic RNA. As output, a damage score ranging from 0 to 1 is given for each cell providing an intuitive scale for filtering that is standardised across cell types, samples, and experiments.">
<meta property="og:description" content="Detects and filters damaged cells in single-cell RNA sequencing (scRNA-seq) data using a novel approach inspired by DoubletFinder. Damage is detected by measuring the extent to which cells deviate from artificially damaged profiles of themselves, simulated through the probabilistic escape of cytoplasmic RNA. As output, a damage score ranging from 0 to 1 is given for each cell providing an intuitive scale for filtering that is standardised across cell types, samples, and experiments.">
<meta property="og:image" content="https://alicenjoyhenning.github.io/DamageDetective/logo.svg">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">DamageDetective</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/detection-vignette.html">Damage cell quality control</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/alicenjoyhenning/DamageDetective/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.svg" class="logo" alt=""><h1 id="damagedetective-">DamageDetective <a class="anchor" aria-label="anchor" href="#damagedetective-"></a>
</h1>
</div>
<!-- badges: start -->

<!-- badges: end -->
<div class="section level2">
<h2 id="content">Content<a class="anchor" aria-label="anchor" href="#content"></a>
</h2>
<p><a href="#description">Description</a> | <a href="#installation">Installation</a> | <a href="#quick-start">Quick start</a> | <a href="#authors">Authors</a> | <a href="#license">License</a> | <a href="#references">References</a></p>
</div>
<div class="section level2">
<h2 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h2>
<p><a href="https://alicenjoyhenning.github.io/DamageDetective/"><strong>View on DamageDetective website</strong></a></p>
<p>Damaged cells are an artifact of single-cell RNA-seq (scRNA-seq) data. These are cells that succumbed to stress before, or in the process of, being sequenced and as a result capture gene expression data that fails to describe the cells in their viable states. As such, these cells are important targets for quality control (QC) filtering.</p>
<p>Current approaches detect damage according to deviation in cell-level QC metrics. These approaches assume all viable cells follow similar distributions across QC metrics, an assumption that does not hold in heterogeneous samples.</p>
<p>More recent approaches address this by analysing cells at a population level, isolating cells with similar distributions before addressing deviations within them. This assumes all distinct distributions are associated with viable populations, putting damaged cells at risk of misclassification if they are abundant. Ultimately, filtering decisions of all current approaches rely more on statistical definitions of deviation than biological definitions of damage.</p>
<p><code>DamageDetective</code> takes a different approach, rather than detecting damage by measuring the extent to which cells deviate from each other, damage is detected by measuring the extent to which cells deviate from artificially damaged profiles of themselves. This is inspired by the approach of <code>DoubletFinder</code><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>1</mn></msup><annotation encoding="application/x-tex">^1</annotation></semantics></math>, a high performing QC tool for filtering doublets, another prominent scRNA-seq artifact. By comparing the expression profiles of true cells to sets of artificially damaged cells, <code>DamageDetective</code> estimates the damage severity of true cells as a score from 0 to 1. This provides an intuitive scale for filtering that is standardised across cell types, samples, and experiments.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><code>DamageDetective</code> can soon be installed from CRAN,</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'DamageDetective'</span><span class="op">)</span></span></code></pre></div>
<p>Or is available directly through GitHub,</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AlicenJoyHenning/DamageDetective"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
<p>To verify installation, run the following to see if you can view the package vignette and the function help pages,</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alicenjoyhenning.github.io/DamageDetective/">DamageDetective</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"DamageDetective"</span><span class="op">)</span></span></code></pre></div>
<p><br><br></p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The demonstrations below can be followed immediately after loading the package and serve as a test to ensure all is running smoothly. For more detailed examples and explanations, please refer to the package articles available on our <a href="https://alicenjoyhenning.github.io/DamageDetective/">website</a>.</p>
<div class="section level3">
<h3 id="prepare-input">Prepare input<a class="anchor" aria-label="anchor" href="#prepare-input"></a>
</h3>
<p>Damage detection is carried out by the <code>detect_damage</code> function that requires a count matrix to run. We will use the dummy count matrix provided by <code>DamageDetective</code>, <code>test_counts</code>, a subset of the <a href="10.1038/s41591-020-0769-8%5D">(kotliarov-pbmc-2020)</a> PBMC dataset provided in the <code>scRNAseq</code><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mn>2</mn></msup><annotation encoding="application/x-tex">^2</annotation></semantics></math> package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alicenjoyhenning.github.io/DamageDetective/">DamageDetective</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"test_counts"</span>, package <span class="op">=</span> <span class="st">"DamageDetective"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">test_counts</span><span class="op">)</span></span>
<span><span class="co"># [1] 32738   500</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="select-parameters-for-damage-detection">Select parameters for damage detection<a class="anchor" aria-label="anchor" href="#select-parameters-for-damage-detection"></a>
</h3>
<div class="section level4">
<h4 id="ribosome_penalty">
<code>ribosome_penalty</code><a class="anchor" aria-label="anchor" href="#ribosome_penalty"></a>
</h4>
<p>While <code>detect_damage</code> requires only a count matrix as input, there are optional parameters that control the implementation of the function. Of these, we recommend <code>ribosome_penalty</code> be adjusted for each dataset. This can be done automatically using <code>select_penalty</code>, a function that also requires only the count matrix as input and will output a numeric of the optimal penalty.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">penalty</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/select_penalty.html">select_penalty</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View penalty</span></span>
<span><span class="va">penalty</span></span>
<span><span class="co"># [1] 1e-05</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filter_threshold">
<code>filter_threshold</code><a class="anchor" aria-label="anchor" href="#filter_threshold"></a>
</h4>
<p><code>DamageDetective</code> does not provide binary classifications, “cell” or “damage”, it provides a score from 0 to 1 indicating the estimated extent of damage in the cell. This is taken directly from the extent of cytoplasmic RNA loss, a proportion ranging from 0 to 1, of the set of artificial cells which a true cell has the greatest proximity to.</p>
<p>In other words, the user must choose a value between 0 and 1 to determine the level of damage, or proportion of RNA loss, above which cells will be excluded. By default, <code>DamageDetective</code> offers the threshold of <code>0.7</code>. Values greater than <code>0.7</code> reflect more permissive filtering while those closer to <code>0</code> reflect more stringent filtering. We recommend the default for all cases but suggest that if adjustments are made, they are informed by inspecting the output <code>detect_damage</code> plots, <code>generate_plot = TRUE</code>.</p>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="run-damage-detection">Run damage detection<a class="anchor" aria-label="anchor" href="#run-damage-detection"></a>
</h3>
<p>Damage detection can be run using the count matrix and selected parameters as inputs. Below we have specified <code>filter_counts</code> parameter to <code>TRUE</code>. This will use the default <code>filter_threshold</code> to identify cells for removal and return the filtered count matrix as output.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform damage detection</span></span>
<span><span class="va">detection_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/detect_damage.html">detect_damage</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>,</span>
<span>  ribosome_penalty <span class="op">=</span> <span class="va">penalty</span>,</span>
<span>  filter_counts <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Simulating cells between 1e-05 and 0.08 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.1 and 0.3 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.3 and 0.5 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.5 and 0.7 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.7 and 0.9 RNA loss...</span></span>
<span><span class="co"># Computing pANN...</span></span>
<span></span>
<span><span class="co"># View the resulting count matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">detection_results</span><span class="op">$</span><span class="va">output</span><span class="op">)</span></span>
<span><span class="co"># [1] 32738   470</span></span></code></pre></div>
<blockquote>
<p>Note, the above assumes the data is of human origin, see <code>organism</code> parameter.</p>
</blockquote>
<div style="text-align: center;">
<div class="float">
<img src="reference/figures/plot.svg" alt="Output of test_counts with cells coloured according to the estimated level of damage"><div class="figcaption">Output of test_counts with cells coloured according to the estimated level of damage</div>
</div>
</div>
<p><br></p>
<p>Alternatively, if <code>filter_counts</code> is set to <code>FALSE</code>, a data frame will be given as output containing the damage scores for each barcode. This is provided for the user if they wish to interact with the <code>DamageDetective</code> results more directly. From here, a user can filter their manually data as done by <code>filter_counts</code> automatically.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform damage detection</span></span>
<span><span class="va">detection_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/detect_damage.html">detect_damage</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>,</span>
<span>  ribosome_penalty <span class="op">=</span> <span class="va">penalty</span>,</span>
<span>  filter_counts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Simulating cells between 1e-05 and 0.08 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.1 and 0.3 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.3 and 0.5 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.5 and 0.7 RNA loss...</span></span>
<span><span class="co"># Simulating cells between 0.7 and 0.9 RNA loss...</span></span>
<span><span class="co"># Computing pANN...</span></span>
<span></span>
<span><span class="co"># View output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">detection_results</span><span class="op">$</span><span class="va">output</span><span class="op">)</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#                     Cells DamageDetective</span></span>
<span><span class="co"># TCTGGAAAGCCCAACC_H1B2ln6      0.03826609</span></span>
<span><span class="co"># CCGTTCATCGTGGGAA_H1B2ln2      0.48333333</span></span>
<span><span class="co"># CTTCTCTTCAGCCTAA_H1B2ln1      0.05217739</span></span>
<span><span class="co"># GGATTACAGGGATGGG_H1B2ln1      0.01044348</span></span>
<span><span class="co"># TCTATTGTCTGGTATG_H1B2ln2      0.02435478</span></span>
<span><span class="co"># ACGGGTCAGACAAGCC_H1B2ln6      0.28888889</span></span>
<span></span>
<span><span class="co"># Filter matrix </span></span>
<span><span class="va">undamaged_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">detection_results</span><span class="op">$</span><span class="va">output</span>, <span class="va">DamageDetective</span> <span class="op">&lt;</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span><span class="va">filtered_matrix</span> <span class="op">&lt;-</span> <span class="va">test_counts</span><span class="op">[</span>, <span class="va">undamaged_cells</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">filtered_matrix</span><span class="op">)</span></span>
<span><span class="co"># [1] 32738   470</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="license">License<a class="anchor" aria-label="anchor" href="#license"></a>
</h2>
<p><code>DamageDetective</code> is made available for public use through the <a href="https://opensource.org/license/agpl-v3" class="external-link">GNU AGPL-3.0</a></p>
</div>
<div class="section level2">
<h2 id="authors">Authors<a class="anchor" aria-label="anchor" href="#authors"></a>
</h2>
<p><strong>Alicen Henning</strong><br>
Stellenbosch University, Cape Town, South Africa<br>
Bioinformatics and Computational Biology</p>
<p>This work was done under the supervision of Prof Marlo Möller, Prof Gian van der Spuy, and Prof André Loxton.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>McGinnis, C. S., Murrow, L. M., &amp; Gartner, Z. J. (2019). DoubletFinder: Doublet Detection in Single-Cell RNA Sequencing Data Using Artificial Nearest Neighbors. <em>Cell Systems, 8</em>(4), 329-337.e4. <a href="https://doi.org/10.1016/j.cels.2019.03.003" class="external-link uri">https://doi.org/10.1016/j.cels.2019.03.003</a></p></li>
<li><p>Risso D, Cole M (2024). <em>scRNAseq: Collection of Public Single-Cell RNA-Seq Datasets</em>. <a href="doi:10.18129/B9.bioc.scRNAseq" class="uri">doi:10.18129/B9.bioc.scRNAseq</a> <a href="https://doi.org/10.18129/B9.bioc.scRNAseq" class="external-link uri">https://doi.org/10.18129/B9.bioc.scRNAseq</a>, R package version 2.20.0, <a href="https://bioconductor.org/packages/scRNAseq" class="external-link uri">https://bioconductor.org/packages/scRNAseq</a>.</p></li>
</ol>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=DamageDetective" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/alicenjoyhenning/DamageDetective/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/alicenjoyhenning/DamageDetective/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>AGPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing DamageDetective</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Alicen Henning <br><small class="roles"> Author, maintainer, copyright holder </small> <a href="https://orcid.org/0009-0004-0535-1624" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cosimameyer/overviewR/actions" class="external-link"><img src="https://github.com/AlicenJoyHenning/DamageDetective/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alicen Henning.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
