---
title: "DamageDetective"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DamageDetective}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 7,
  fig.height = 7
)
```

```{r setup}
library(DamageDetective)
```

The demonstrations below can be followed immediately after installing the package by loading the dummy data provided, `test_counts`, an artificial PBMC dataset containing 500 cells and 10009 genes.

## Simulating Damaged Cells (`simulate_counts`)

A core feature of `DamageDetective` is the ability to predict how scRNA-seq data might appear if cells had undergone RNA damage. This is modeled by simulating the loss of cytoplasmic RNA, where greater loss corresponds to increased damage severity.

The `simulate_counts` function applies this simulation, allowing users to introduce damage to a defined proportion of cells in an input count matrix. While primarily used within `detect_damage`, it is also available for standalone use to aid exploratory data analysis and ground truth dataset generation.

```{r}
# Simulating damage in 25% of cells
damaged_counts <- simulate_counts(
  count_matrix = test_counts,
  damage_proportion = 0.20
)

# View dimensions
dim(damaged_counts$matrix)

# Preview the altered counts
damaged_counts$matrix[1:5, 1:5]
```

The function also outputs a `qc_summary` table comparing quality control metrics before and after simulation:

```{r}
head(damaged_counts$qc_summary, 5)
```

Additionally, `simulate_counts` produces plots illustrating quality metric distributions before and after simulated damage that are displayed by default and accessed as shown below.

```{r}
damaged_counts$plot
```

To control the severity and distribution of simulated damage, users can adjust:   
- `target_damage`: Specifies the range of RNA loss per cell.   
- `damage_distribution`: Defines how damage is spread across cells.

For example, simulating severe damage (70-99% RNA loss) with a right-skewed distribution:

```{r}
damaged_counts <- simulate_counts(
  count_matrix = test_counts,
  damage_proportion = 0.25,
  target_damage = c(0.7, 0.99),
  damage_distribution = "right_skewed"
)
```

For more details, refer to:

```{r}
# ?simulate_counts()
```


## Detecting Damaged Cells (`detect_damage`)

`DamageDetective` offers automated damage assessment via the `detect_damage` function, which assigns each cell a damage score.

Before using `detect_damage`, it is important to ensure damage simulation is as accurate as possible for your data. Based on observations of true single cell data, we find that ribosomal RNA loss occurs less frequently than expected based on abundance alone and needs to be adjusted by a numerical value (`ribosome_penalty`) between 0 and 1.

While lower values for `ribosome_penalty` (closer to zero) better approximate true data, with a default of 0.01, this can be refined for the input data using the `select_penalty` function.

```{r}
penalty <- select_penalty(
  count_matrix = test_counts
)

print(penalty)
```

For more details, refer to:

```{r}
# ?select_penalty()
```

Using the refined ribosomal penalty found, the damage detection can then be run as follows,

```{r, fig.width=7, fig.height=4}

damage_results <- detect_damage(
  count_matrix = test_counts, 
  ribosome_penalty = penalty
)

# View the damage scores
head(damage_results, 5)

```

Cells with high damage scores can be filtered out to retain high-quality data:

```{r}
filtered_counts <- test_counts[, damage_results$qc_summary$DamageDetective < 0.7]
```

For convenience, `detect_damage` can return a filtered count matrix directly by enabling `filter_counts = TRUE`.

```{r}
filtered_counts <- detect_damage(
  count_matrix = test_counts,
  ribosome_penalty = 0.01,
  filter_counts = TRUE,
  filter_threshold = 0.7, 
  generate_plot = FALSE
)

class(filtered_counts)
dim(filtered_counts)
```

For more details, refer to:

```{r}
# ?detect_damage()
```
