<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>detect_damage — detect_damage • DamageDetective</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="detect_damage — detect_damage"><meta name="description" content="Quality control function to identify and filter damaged cells from an
input count matrix, where 'damage' is defined by the loss of cytoplasmic RNA."><meta property="og:description" content="Quality control function to identify and filter damaged cells from an
input count matrix, where 'damage' is defined by the loss of cytoplasmic RNA."><meta property="og:image" content="https://alicenjoyhenning.github.io/DamageDetective/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DamageDetective</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/detection-vignette.html">Damage cell quality control</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/alicenjoyhenning/DamageDetective/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>detect_damage</h1>
      <small class="dont-index">Source: <a href="https://github.com/alicenjoyhenning/DamageDetective/blob/main/R/detect_damage.R" class="external-link"><code>R/detect_damage.R</code></a></small>
      <div class="d-none name"><code>detect_damage.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Quality control function to identify and filter damaged cells from an
input count matrix, where 'damage' is defined by the loss of cytoplasmic RNA.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">detect_damage</span><span class="op">(</span></span>
<span>  <span class="va">count_matrix</span>,</span>
<span>  ribosome_penalty <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Hsap"</span>,</span>
<span>  annotated_celltypes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  target_damage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.65</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  damage_distribution <span class="op">=</span> <span class="st">"right_skewed"</span>,</span>
<span>  distribution_steepness <span class="op">=</span> <span class="st">"moderate"</span>,</span>
<span>  pca_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log.features"</span>, <span class="st">"log.counts"</span>, <span class="st">"mt.prop"</span>, <span class="st">"rb.prop"</span><span class="op">)</span>,</span>
<span>  beta_shape_parameters <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  damage_proportion <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  kN <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  generate_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  display_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"#7023FD"</span>, <span class="st">"#E60006"</span><span class="op">)</span>,</span>
<span>  filter_threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  filter_counts <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-count-matrix">count_matrix<a class="anchor" aria-label="anchor" href="#arg-count-matrix"></a></dt>
<dd><p>Matrix or dgCMatrix containing the counts from
single cell RNA sequencing data.</p></dd>


<dt id="arg-ribosome-penalty">ribosome_penalty<a class="anchor" aria-label="anchor" href="#arg-ribosome-penalty"></a></dt>
<dd><p>Numeric specifying the factor by which the
probability of loosing a transcript from a ribosomal gene is multiplied by.
Here, values closer to 0 represent a greater penalty.</p><ul><li><p>Default is 0.01.</p></li>
</ul></dd>


<dt id="arg-organism">organism<a class="anchor" aria-label="anchor" href="#arg-organism"></a></dt>
<dd><p>String specifying the organism of origin of the input
data where there are two standard options,</p><ul><li><p>"Hsap"</p></li>
<li><p>"Mmus"</p></li>
</ul><p>If a user wishes to use a non-standard organism they must input a list
containing strings for the patterns to match mitochondrial and ribosomal
genes of the organism. If available, nuclear-encoded genes that are likely
retained in the nucleus, such as in nuclear speckles, must also
be specified. An example for humans is below,</p><ul><li><p>organism = list(mito_pattern = "^MT-",
ribo_pattern = "^(RPS|RPL)",
nuclear = c("NEAT1","XIST", "MALAT1"))</p></li>
<li><p>Default is "Hsap"</p></li>
</ul></dd>


<dt id="arg-annotated-celltypes">annotated_celltypes<a class="anchor" aria-label="anchor" href="#arg-annotated-celltypes"></a></dt>
<dd><p>Boolean specifying whether input matrix has
cell type information stored.</p><ul><li><p>Default is FALSE</p></li>
</ul></dd>


<dt id="arg-target-damage">target_damage<a class="anchor" aria-label="anchor" href="#arg-target-damage"></a></dt>
<dd><p>Numeric vector specifying the upper and lower range of
the level of damage that will be introduced.</p>
<p>Here, damage refers to the amount of cytoplasmic RNA lost by a cell where
values closer to 1 indicate more loss and therefore more heavily damaged
cells.</p><ul><li><p>Default is c(0.1, 0.8)</p></li>
</ul></dd>


<dt id="arg-damage-distribution">damage_distribution<a class="anchor" aria-label="anchor" href="#arg-damage-distribution"></a></dt>
<dd><p>String specifying whether the distribution of
damage levels among the damaged cells should be shifted towards the
upper or lower range of damage specified in 'target_damage' or follow
a symmetric distribution between them. There are three valid options:</p><ul><li><p>"right_skewed"</p></li>
<li><p>"left_skewed"</p></li>
<li><p>"symmetric"</p></li>
<li><p>Default is "right_skewed"</p></li>
</ul></dd>


<dt id="arg-distribution-steepness">distribution_steepness<a class="anchor" aria-label="anchor" href="#arg-distribution-steepness"></a></dt>
<dd><p>String specifying how concentrated the spread
of damaged cells are about the mean of the target distribution specified in
'target_damage'. Here, an increase in steepness manifests in a more
apparent skewness.There are three valid options:</p><ul><li><p>"shallow"</p></li>
<li><p>"moderate"</p></li>
<li><p>"steep"</p></li>
<li><p>Default is "moderate"</p></li>
</ul></dd>


<dt id="arg-pca-columns">pca_columns<a class="anchor" aria-label="anchor" href="#arg-pca-columns"></a></dt>
<dd><p>String vector containing the names of the metrics used
for principal component analysis.</p>
<p>There are 8 options:</p><ul><li><p>features</p></li>
<li><p>log.features</p></li>
<li><p>counts</p></li>
<li><p>log.counts</p></li>
<li><p>mt.prop</p></li>
<li><p>rb.prop</p></li>
<li><p>malat1.prop</p></li>
<li><p>malat1</p></li>
</ul><p>Default is c("log.features", "log.counts", "mt.prop", "rb.prop").</p></dd>


<dt id="arg-beta-shape-parameters">beta_shape_parameters<a class="anchor" aria-label="anchor" href="#arg-beta-shape-parameters"></a></dt>
<dd><p>Numeric vector that allows for the shape
parameters of the beta distribution to defined explicitly. This offers
greater flexibility than allowed by the 'damage_distribution' and
'distribution_steepness' parameters and will override the defaults they
offer.</p><ul><li><p>Default is 'NULL'</p></li>
</ul></dd>


<dt id="arg-damage-proportion">damage_proportion<a class="anchor" aria-label="anchor" href="#arg-damage-proportion"></a></dt>
<dd><p>Numeric describing what proportion
of the input data should be altered to resemble damaged data.</p><ul><li><p>Must range between 0 and 1.</p></li>
</ul></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Numeric specifying the random seed to ensure reproducibility of
the function's output. Setting a seed ensures that the random sampling
and perturbation processes produce the same results when the function
is run multiple times with the same input data and parameters.</p><ul><li><p>Default is 7.</p></li>
</ul></dd>


<dt id="arg-kn">kN<a class="anchor" aria-label="anchor" href="#arg-kn"></a></dt>
<dd><p>Numeric describing how many nearest neighbours are considered
for pANN calculations. kN cannot exceed the total cell number.</p><ul><li><p>Default is one third of the total cell number.</p></li>
</ul></dd>


<dt id="arg-generate-plot">generate_plot<a class="anchor" aria-label="anchor" href="#arg-generate-plot"></a></dt>
<dd><p>Boolean specifying whether the QC plot should
be outputted. QC plots will be generated by default as we recommend
verifying the perturbed data retains characteristics of true
single cell data.</p><ul><li><p>Default is TRUE.</p></li>
</ul></dd>


<dt id="arg-display-plot">display_plot<a class="anchor" aria-label="anchor" href="#arg-display-plot"></a></dt>
<dd><p>Boolean specifying whether the output QC plot should
be displayed in the global environment. Naturally, this is only relevant
when generate_plot is TRUE.</p><ul><li><p>Default is TRUE.</p></li>
</ul></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>String specifying the three colours that will be used to
create the continuous colour palette for colouring the 'damage_column'.</p><ul><li><p>Default is a range from purple to red,
c("grey", "#7023FD", "#E60006").</p></li>
</ul></dd>


<dt id="arg-filter-threshold">filter_threshold<a class="anchor" aria-label="anchor" href="#arg-filter-threshold"></a></dt>
<dd><p>Numeric specifying the proportion of RNA loss
above which a cell should be considered damaged.</p><ul><li><p>Default is 0.5.</p></li>
</ul></dd>


<dt id="arg-filter-counts">filter_counts<a class="anchor" aria-label="anchor" href="#arg-filter-counts"></a></dt>
<dd><p>Boolean specifying whether the output matrix
should be filtered, returned containing only cells that fall below
the filter threshold. Alternatively, a data frame containing cell
barcodes and their associated label as either 'damaged' or 'cell'
is returned.</p><ul><li><p>Default is FALSE.</p></li>
</ul></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Boolean specifying whether messages and function progress
should be displayed in the console.</p><ul><li><p>Default is TRUE.</p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Filtered matrix or data frame containing damage labels.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Using the simulation framework of <code><a href="simulate_counts.html">simulate_counts()</a></code>, <code>detect_damage()</code>
generates artificially damaged cell profiles by introducing defined levels
of RNA loss into the input data. True and artificial cells are then
merged and pre-processed to compute the following quality control metrics:</p><ul><li><p>Log-normalized feature count</p></li>
<li><p>Log-normalized total counts</p></li>
<li><p>Mitochondrial proportion</p></li>
<li><p>Ribosomal proportion</p></li>
<li><p>Log-normalized MALAT1 gene expression</p></li>
</ul><p>Principal component analysis (PCA) is performed on these metrics,
and a Euclidean distance matrix is constructed from the PC embeddings.
For each true cell, the proportion of nearest neighbours that are
artificial cells (pANN) is calculated across all damage levels and the
damage level with the highest pANN is assigned to the true cell.
Finally, cells exceeding a specified damage threshold, <code>filter_threshold</code>,
are marked as damaged.</p>
<p>This filtering method is inspired by approaches developed for DoubletFinder
(McGinnis et al., 2019) to detect doublets in single-cell data.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>McGinnis, C. S., Murrow, L. M., &amp; Gartner, Z. J. (2019). DoubletFinder:
Doublet Detection in Single-Cell RNA Sequencing Data Using Artificial
Nearest neighbours. <em>Cell Systems, 8</em>(4), 329-337.e4.
<a href="https://doi.org/10.1016/j.cels.2019.03.003" class="external-link">doi:10.1016/j.cels.2019.03.003</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"test_counts"</span>, package <span class="op">=</span> <span class="st">"DamageDetective"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">test_counts</span> <span class="op">&lt;-</span> <span class="va">test_counts</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu">detect_damage</span><span class="op">(</span></span></span>
<span class="r-in"><span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>,</span></span>
<span class="r-in"><span>  generate_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>  seed <span class="op">=</span> <span class="fl">7</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Simulating damage...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> PCA with log.features, mt.prop, malat1.prop</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing pANN...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alicen Henning.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

