<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>simulate_counts — simulate_counts • DamageDetective</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="simulate_counts — simulate_counts"><meta name="description" content="Function to simulate damaged cells by perturbing the gene expression of
existing cells."><meta property="og:description" content="Function to simulate damaged cells by perturbing the gene expression of
existing cells."><meta property="og:image" content="https://alicenjoyhenning.github.io/DamageDetective/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DamageDetective</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.15</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/alicenjoyhenning/DamageDetective/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>simulate_counts</h1>
      <small class="dont-index">Source: <a href="https://github.com/alicenjoyhenning/DamageDetective/blob/main/R/simulate_counts.R" class="external-link"><code>R/simulate_counts.R</code></a></small>
      <div class="d-none name"><code>simulate_counts.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Function to simulate damaged cells by perturbing the gene expression of
existing cells.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">simulate_counts</span><span class="op">(</span></span>
<span>  <span class="va">count_matrix</span>,</span>
<span>  damage_proportion <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  annotated_celltypes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  target_damage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.65</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  damage_distribution <span class="op">=</span> <span class="st">"left_skewed"</span>,</span>
<span>  distribution_steepness <span class="op">=</span> <span class="st">"steep"</span>,</span>
<span>  beta_shape_parameters <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ribosome_penalty <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  generate_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"#7023FD"</span>, <span class="st">"#E60006"</span><span class="op">)</span>,</span>
<span>  plot_ribosomal_penalty <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  display_plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Hsap"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-count-matrix">count_matrix<a class="anchor" aria-label="anchor" href="#arg-count-matrix"></a></dt>
<dd><p>Matrix or dgCMatrix containing the counts from
single cell RNA sequencing data.</p></dd>


<dt id="arg-damage-proportion">damage_proportion<a class="anchor" aria-label="anchor" href="#arg-damage-proportion"></a></dt>
<dd><p>Numeric describing what proportion
of the input data should be altered to resemble damaged data.</p><ul><li><p>Must range between 0 and 1.</p></li>
</ul></dd>


<dt id="arg-annotated-celltypes">annotated_celltypes<a class="anchor" aria-label="anchor" href="#arg-annotated-celltypes"></a></dt>
<dd><p>Boolean specifying whether input matrix has
cell type information stored.</p><ul><li><p>Default is FALSE</p></li>
</ul></dd>


<dt id="arg-target-damage">target_damage<a class="anchor" aria-label="anchor" href="#arg-target-damage"></a></dt>
<dd><p>Numeric vector specifying the upper and lower range of
the level of damage that will be introduced.</p>
<p>Here, damage refers to the amount of cytoplasmic RNA lost by a cell where
values closer to 1 indicate more loss and therefore more heavily damaged
cells.</p><ul><li><p>Default is c(0.65, 1)</p></li>
</ul></dd>


<dt id="arg-damage-distribution">damage_distribution<a class="anchor" aria-label="anchor" href="#arg-damage-distribution"></a></dt>
<dd><p>String specifying whether the distribution of
damage levels among the damaged cells should be shifted towards the
upper or lower range of damage specified in 'target_damage' or follow
a symmetric distribution between them. There are three valid options:</p><ul><li><p>"right_skewed"</p></li>
<li><p>"left_skewed"</p></li>
<li><p>"symmetric"</p></li>
<li><p>Default is "left_skewed"</p></li>
</ul></dd>


<dt id="arg-distribution-steepness">distribution_steepness<a class="anchor" aria-label="anchor" href="#arg-distribution-steepness"></a></dt>
<dd><p>String specifying how concentrated the spread
of damaged cells are about the mean of the target distribution specified in
'target_damage'. Here, an increase in steepness manifests in a more
apparent skewness.There are three valid options:</p><ul><li><p>"shallow"</p></li>
<li><p>"moderate"</p></li>
<li><p>"steep"</p></li>
<li><p>Default is "steep"</p></li>
</ul></dd>


<dt id="arg-beta-shape-parameters">beta_shape_parameters<a class="anchor" aria-label="anchor" href="#arg-beta-shape-parameters"></a></dt>
<dd><p>Numeric vector that allows for the shape
parameters of the beta distribution to defined explicitly. This offers
greater flexibility than allowed by the 'damage_distribution' and
'distribution_steepness' parameters and will override the defaults they
offer.</p><ul><li><p>Default is 'NULL'</p></li>
</ul></dd>


<dt id="arg-ribosome-penalty">ribosome_penalty<a class="anchor" aria-label="anchor" href="#arg-ribosome-penalty"></a></dt>
<dd><p>Numeric specifying the factor by which the
probability of loosing a transcript from a ribosomal gene is multiplied by.
Here, values closer to 0 represent a greater penalty.</p><ul><li><p>Default is 1.</p></li>
</ul></dd>


<dt id="arg-generate-plot">generate_plot<a class="anchor" aria-label="anchor" href="#arg-generate-plot"></a></dt>
<dd><p>Boolean specifying whether the QC plot should
be outputted. QC plots will be generated by default as we recommend
verifying the perturbed data retains characteristics of true
single cell data.</p><ul><li><p>Default is TRUE.</p></li>
</ul></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>Character vector containing three colours to create the
continuous palette for damaged cells.</p><ul><li><p>Default is c("grey", "#7023FD", "#E60006").</p></li>
</ul></dd>


<dt id="arg-plot-ribosomal-penalty">plot_ribosomal_penalty<a class="anchor" aria-label="anchor" href="#arg-plot-ribosomal-penalty"></a></dt>
<dd><p>Boolean specifying whether the output QC plot
should focus on only the ribosomal proportion or contain additional QC
information. If TRUE, this can be useful for visualising the impact of
the ribosomal penalty parameter.</p><ul><li><p>Default is FALSE.</p></li>
</ul></dd>


<dt id="arg-display-plot">display_plot<a class="anchor" aria-label="anchor" href="#arg-display-plot"></a></dt>
<dd><p>Boolean specifying whether the output QC plot should
be displayed in the global environment. Naturally, this is only relevant
when generate_plot is TRUE.</p><ul><li><p>Default is TRUE.</p></li>
</ul></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Numeric specifying the random seed to ensure reproducibility of
the function's output. Setting a seed ensures that the random sampling
and perturbation processes produce the same results when the function
is run multiple times with the same input data and parameters.</p><ul><li><p>Default is 7.</p></li>
</ul></dd>


<dt id="arg-organism">organism<a class="anchor" aria-label="anchor" href="#arg-organism"></a></dt>
<dd><p>String specifying the organism of origin of the input
data where there are two standard options,</p><ul><li><p>"Hsap"</p></li>
<li><p>"Mmus"</p></li>
</ul><p>If a user wishes to use a non-standard organism they must input a list
containing strings for the patterns to match mitochondrial and ribosomal
genes of the organism. If available, nuclear-encoded genes that are likely
retained in the nucleus, such as in nuclear speckles, must also
be specified. An example for humans is below,</p><ul><li><p>organism = list(mito_pattern = "^MT-",
ribo_pattern = "^(RPS|RPL)",
nuclear = c("NEAT1","XIST", "MALAT1"))</p></li>
<li><p>Default is "Hsap"</p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the altered count matrix, a data frame with summary
statistics, and, if specified, a 'ggplot2' object of the quality control
metrics of the alteration.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>'DamageDetective' models damage in single-cell RNA sequencing data as the
loss of cytoplasmic RNA, where cells experiencing greater RNA loss are
assumed to be more extensively damaged, while those with minimal loss are
considered largely intact. The perturbation process introduces RNA loss into
existing cells and is controlled by three key parameters: the <strong>target
proportion of damage</strong>,  which specifies the fraction of cells to be
perturbed; the <strong>target level of damage</strong>, which defines the extent of RNA
loss across cells; and the <strong>target distribution of damage</strong>, which
determines how the different levels of RNA loss are distributed across
cells.</p>
<p>Based on these parameters, cells are randomly selected and assigned a target
proportion of RNA loss. The total number of transcripts to be removed is
determined, and perturbation is applied through weighted sampling without
replacement from cytoplasmic gene counts. Here, the probability of
transcript loss is determined by gene abundance, with highly expressed genes
more likely to lose transcripts. Once the target RNA loss is reached, the
cell's expression profile is updated, and the process repeats for all
selected cells.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"test_counts"</span>, package <span class="op">=</span> <span class="st">"DamageDetective"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">simulated_damage</span> <span class="op">&lt;-</span> <span class="fu">simulate_counts</span><span class="op">(</span></span></span>
<span class="r-in"><span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>,</span></span>
<span class="r-in"><span>  damage_proportion <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>  ribosome_penalty <span class="op">=</span> <span class="fl">0.25</span>,</span></span>
<span class="r-in"><span>  target_damage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  generate_plot <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alicen Henning.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

