<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>DamageDetective Overview • DamageDetective</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DamageDetective Overview">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DamageDetective</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.15</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/alicenjoyhenning/DamageDetective/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>DamageDetective Overview</h1>
            
            <h4 data-toc-skip class="date">Compiled: July 16, 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/alicenjoyhenning/DamageDetective/blob/main/vignettes/detection-vignette.Rmd" class="external-link"><code>vignettes/detection-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>detection-vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="package-description">Package description<a class="anchor" aria-label="anchor" href="#package-description"></a>
</h2>
<p>The goal of <code>DamageDetective</code> is to simplify the process
of making informed and reproducible damaged cell filtering decisions
during the pre-processing of single cell RNA sequencing data.</p>
<p>This requires only a count matrix to run and outputs a damage score
ranging from 0 (viable, intact cell) to 1 (broken, non-viable cell).
There is an option for automatic filtering using the default upper
threshold damage score of 0.5.</p>
<p>We will demonstrate briefly using an example dataset provided upon
package installation.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alicenjoyhenning.github.io/DamageDetective/">DamageDetective</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org" class="external-link">Matrix</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"test_counts"</span>, package <span class="op">=</span> <span class="st">"DamageDetective"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">test_counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 32738   500</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<div class="section level3">
<h3 id="libraries">Libraries<a class="anchor" aria-label="anchor" href="#libraries"></a>
</h3>
<p>For an improvement in speed, load the <code>presto</code> package in
addition to <code>DamageDetective</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"madsen-lab/valiDrops"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://alicenjoyhenning.github.io/DamageDetective/">DamageDetective</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">presto</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-data-formatting">Input data formatting<a class="anchor" aria-label="anchor" href="#input-data-formatting"></a>
</h3>
<ul>
<li>Counts should be provided in the form of a compressed,
column-oriented sparse matrix (<code>dgCMatrix</code>) in
<code>R</code>.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View formatting</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">test_counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "dgCMatrix"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "Matrix"</span></span></code></pre></div>
<ul>
<li>Counts should have gene sets formatted according to HGCN standard,
i.e., <code>MT-...</code> rather than <code>ENSG...</code>.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View formatting</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">test_counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "MIR1302-10"   "FAM138A"      "OR4F5"        "RP11-34P13.7" "RP11-34P13.8"</span></span>
<span><span class="co">#&gt; [6] "AL627309.1"</span></span></code></pre></div>
<pre><code>&gt; See [biomaRt](https://bioconductor.org/packages/release/bioc/html/biomaRt.html) for conversion assistance or use the automated `Seurat` functions for working with alignment output, [ReadMtx](https://satijalab.org/seurat/reference/readmtx).</code></pre>
<ul>
<li>
<code>DamageDetective</code> supports data of human and mouse,
specified using the <code>organism</code> parameter. To analyse a
non-standard organism, provide a list with patterns that matches the set
of mitochondrially encoded genes and ribosomal genes and genes with a
confirmed, permanent nuclear residence.</li>
</ul>
<blockquote>
<p><strong>Example using humans as organism of interest</strong></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">organism</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>mito_pattern <span class="op">=</span> <span class="st">"^MT-"</span>,</span>
<span>                ribo_pattern <span class="op">=</span> <span class="st">"^(RPS|RPL)"</span>,</span>
<span>                nuclear <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NEAT1"</span>,<span class="st">"XIST"</span>, <span class="st">"MALAT1"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</blockquote>
<p>For more information on data preparation, view the package articles
on our <a href="https://alicenjoyhenning.github.io/DamageDetective">website</a></p>
</div>
<div class="section level3">
<h3 id="parameter-selection">Parameter selection<a class="anchor" aria-label="anchor" href="#parameter-selection"></a>
</h3>
<div class="section level4">
<h4 id="select_penalty">
<code>select_penalty</code><a class="anchor" aria-label="anchor" href="#select_penalty"></a>
</h4>
<p>While <code>detect_damage</code> requires only a count matrix as
input, additional parameters control aspects of the function’s
computations. Of these, we recommend <code>ribosome_penalty</code> be
adjusted for each dataset using the <code>select_penalty</code>
function.</p>
<p>This parameter ranges from 0 to 1 and adjusts the likelihood of
ribosomal RNA loss during simulation, correcting for observed
discrepancies where ribosomal RNA is retained more than expected based
on transcript abundance.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">penalty</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_penalty.html">select_penalty</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>, </span>
<span>  max_penalty_trials <span class="op">=</span> <span class="fl">3</span> <span class="co"># Shortened for the vignette</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Testing penalty of 0.1...</span></span>
<span><span class="co">#&gt; Testing penalty of 0.15...</span></span>
<span><span class="co">#&gt; Testing penalty of 0.2...</span></span>
<span><span class="co">#&gt; Maximum penalty trials reached (3). Stopping.</span></span>
<span><span class="va">penalty</span></span>
<span><span class="co">#&gt; [1] 0.2</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="filter_threshold">
<code>filter_threshold</code><a class="anchor" aria-label="anchor" href="#filter_threshold"></a>
</h4>
<p><code>DamageDetective</code> offers the upper threshold 0.5 as the
damage score above which cells are filtered, where values greater than
0.5 reflect more permissive filtering and values closer to 0 reflect
more stringent filtering. We recommend the default, but suggest that if
adjustments are made, they are informed by the output detect_damage
plots, generate_plot = TRUE.</p>
<p>For more information on parameters, please view the function
documentation available on our website under
<strong>References</strong>.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="running-damaged-cell-detection">Running damaged cell detection<a class="anchor" aria-label="anchor" href="#running-damaged-cell-detection"></a>
</h2>
<p>Damage detection is run using the count matrix and ribosomal penalty
as inputs. Below, we have additionally specified for
<code>filter_counts</code> parameter to be TRUE. This will use the
default <code>filter_threshold</code> and return the filtered count
matrix that can be used immediately for the remainder of
pre-processing.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Perform damage detection</span></span>
<span><span class="va">detection_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detect_damage.html">detect_damage</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">test_counts</span>,</span>
<span>  ribosome_penalty <span class="op">=</span> <span class="va">penalty</span>,</span>
<span>  display_plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  filter_counts <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Clustering cells...</span></span>
<span><span class="co">#&gt; For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span></span>
<span><span class="co">#&gt; (default method for FindMarkers) please install the presto package</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; install.packages('devtools')</span></span>
<span><span class="co">#&gt; devtools::install_github('immunogenomics/presto')</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; After installation of presto, Seurat will automatically use the more </span></span>
<span><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; Simulating damage...</span></span>
<span><span class="co">#&gt; Computing pANN...</span></span>
<span></span>
<span><span class="co"># View the resulting count matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">detection_results</span><span class="op">$</span><span class="va">output</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 32738   459</span></span>
<span></span>
<span><span class="co"># View the plot </span></span>
<span><span class="va">detection_results</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="detection-vignette_files/figure-html/damage_detection-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] future_1.58.0          Matrix_1.7-3           DamageDetective_2.0.15</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     jsonlite_2.0.0         magrittr_2.0.3        </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.1-4   farver_2.1.2           rmarkdown_2.29        </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.6               ragg_1.4.0             vctrs_0.6.5           </span></span>
<span><span class="co">#&gt;  [10] ROCR_1.0-11            spatstat.explore_3.4-3 rstatix_0.7.2         </span></span>
<span><span class="co">#&gt;  [13] htmltools_0.5.8.1      broom_1.0.8            Formula_1.2-5         </span></span>
<span><span class="co">#&gt;  [16] sass_0.4.10            sctransform_0.4.2      parallelly_1.45.0     </span></span>
<span><span class="co">#&gt;  [19] KernSmooth_2.23-26     bslib_0.9.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">#&gt;  [22] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">#&gt;  [25] plotly_4.11.0          zoo_1.8-14             cachem_1.1.0          </span></span>
<span><span class="co">#&gt;  [28] igraph_2.1.4           mime_0.13              lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt;  [31] pkgconfig_2.0.3        R6_2.6.1               fastmap_1.2.0         </span></span>
<span><span class="co">#&gt;  [34] fitdistrplus_1.2-4     shiny_1.11.1           digest_0.6.37         </span></span>
<span><span class="co">#&gt;  [37] patchwork_1.3.1        Seurat_5.3.0           tensor_1.5.1          </span></span>
<span><span class="co">#&gt;  [40] RSpectra_0.16-2        irlba_2.3.5.1          textshaping_1.0.1     </span></span>
<span><span class="co">#&gt;  [43] ggpubr_0.6.1           labeling_0.4.3         progressr_0.15.1      </span></span>
<span><span class="co">#&gt;  [46] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-7       </span></span>
<span><span class="co">#&gt;  [49] abind_1.4-8            compiler_4.5.1         proxy_0.4-27          </span></span>
<span><span class="co">#&gt;  [52] withr_3.0.2            backports_1.5.0        carData_3.0-5         </span></span>
<span><span class="co">#&gt;  [55] fastDummies_1.7.5      ggsignif_0.6.4         MASS_7.3-65           </span></span>
<span><span class="co">#&gt;  [58] tools_4.5.1            lmtest_0.9-40          httpuv_1.6.16         </span></span>
<span><span class="co">#&gt;  [61] future.apply_1.20.0    goftest_1.2-3          glue_1.8.0            </span></span>
<span><span class="co">#&gt;  [64] nlme_3.1-168           promises_1.3.3         grid_4.5.1            </span></span>
<span><span class="co">#&gt;  [67] Rtsne_0.17             cluster_2.1.8.1        reshape2_1.4.4        </span></span>
<span><span class="co">#&gt;  [70] generics_0.1.4         gtable_0.3.6           spatstat.data_3.1-6   </span></span>
<span><span class="co">#&gt;  [73] class_7.3-23           tidyr_1.3.1            data.table_1.17.8     </span></span>
<span><span class="co">#&gt;  [76] sp_2.2-0               car_3.1-3              spatstat.geom_3.4-1   </span></span>
<span><span class="co">#&gt;  [79] RcppAnnoy_0.0.22       ggrepel_0.9.6          RANN_2.6.2            </span></span>
<span><span class="co">#&gt;  [82] pillar_1.11.0          stringr_1.5.1          spam_2.11-1           </span></span>
<span><span class="co">#&gt;  [85] RcppHNSW_0.6.0         later_1.4.2            splines_4.5.1         </span></span>
<span><span class="co">#&gt;  [88] dplyr_1.1.4            lattice_0.22-7         survival_3.8-3        </span></span>
<span><span class="co">#&gt;  [91] deldir_2.0-4           tidyselect_1.2.1       miniUI_0.1.2          </span></span>
<span><span class="co">#&gt;  [94] pbapply_1.7-2          knitr_1.50             gridExtra_2.3         </span></span>
<span><span class="co">#&gt;  [97] scattermore_1.2        xfun_0.52              matrixStats_1.5.0     </span></span>
<span><span class="co">#&gt; [100] stringi_1.8.7          lazyeval_0.2.2         yaml_2.3.10           </span></span>
<span><span class="co">#&gt; [103] evaluate_1.0.4         codetools_0.2-20       tibble_3.3.0          </span></span>
<span><span class="co">#&gt; [106] cli_3.6.5              uwot_0.2.3             xtable_1.8-4          </span></span>
<span><span class="co">#&gt; [109] reticulate_1.42.0      systemfonts_1.2.3      jquerylib_0.1.4       </span></span>
<span><span class="co">#&gt; [112] Rcpp_1.1.0             globals_0.18.0         spatstat.random_3.4-1 </span></span>
<span><span class="co">#&gt; [115] png_0.1-8              spatstat.univar_3.1-4  parallel_4.5.1        </span></span>
<span><span class="co">#&gt; [118] pkgdown_2.1.3          ggplot2_3.5.2          dotCall64_1.2         </span></span>
<span><span class="co">#&gt; [121] listenv_0.9.1          viridisLite_0.4.2      scales_1.4.0          </span></span>
<span><span class="co">#&gt; [124] ggridges_0.5.6         e1071_1.7-16           SeuratObject_5.1.0    </span></span>
<span><span class="co">#&gt; [127] purrr_1.1.0            rlang_1.1.6            cowplot_1.2.0</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alicen Henning.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
