<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Damage cell quality control • DamageDetective</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Damage cell quality control">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DamageDetective</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/detection-vignette.html">Damage cell quality control</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/alicenjoyhenning/DamageDetective/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Damage cell quality control</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/alicenjoyhenning/DamageDetective/blob/v1.0.0/vignettes/detection-vignette.Rmd" class="external-link"><code>vignettes/detection-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>detection-vignette.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="tutorial-overview">Tutorial Overview<a class="anchor" aria-label="anchor" href="#tutorial-overview"></a>
</h2>
<p>The goal of <code>DamageDetective</code> is to simplify the process
of making informed and reproducible damaged cell filtering decisions.
This tutorial provides an outline of the steps needed to achieve
this:</p>
<ol style="list-style-type: decimal">
<li>Data preparation</li>
<li>Parameter selection</li>
<li>Damaged cell detection</li>
</ol>
<p>Ensure the following packages are installed and made available in
your R environment to follow along:</p>
<ul>
<li>
<code>DamageDetective</code>, <code>ggplot2</code>,
<code>Seurat</code>, <code>SeuratData</code>, <code>scRNAseq</code>,
<code>SingleCellExperiment</code>, and <code>patchwork</code>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="data-preparation">Data Preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h2>
<p>This tutorial takes place in <code>R</code>, home to three popular
scRNA-seq data storage types:</p>
<ol style="list-style-type: decimal">
<li>The <code>Seurat</code> object</li>
<li>The <code>SingleCellExperiment</code> (<code>sce</code>) object</li>
<li>The matrix retrieved directly from alignment output
(<code>matrix.mtx</code>, <code>barcodes.tsv</code>, and
<code>features.tsv</code>).</li>
</ol>
<p><code>DamageDetective</code> operates with data in a compressed,
column-oriented sparse matrix (<code>dgCMatrix</code>). This format
efficiently handles the sparse nature of single-cell data by indexing
only non-zero elements within each column. Each of the above data
storage types can act as the starting point for damage detection through
conversion to a sparse matrix.</p>
<blockquote>
<p><strong>Note:</strong> Due to vignette size constraints, this
tutorial will only demonstrate live using data from the alignment
output.</p>
</blockquote>
<div class="section level3">
<h3 id="conversion-to-sparse-matrix">Conversion to Sparse Matrix<a class="anchor" aria-label="anchor" href="#conversion-to-sparse-matrix"></a>
</h3>
<div class="section level4">
<h4 id="using-a-seurat-object">Using a <code>Seurat</code> Object<a class="anchor" aria-label="anchor" href="#using-a-seurat-object"></a>
</h4>
<p>The count matrix can be extracted from the assay slot of a Seurat
object. By default, this matrix is already in compressed column-oriented
sparse form and can be used as input for DamageDetective.</p>
<p>We will use the SeuratData package to retrieve a publicly available
Seurat object to demonstrate. This object stores the data from
peripheral blood mononuclear cells (PBMCs) sequenced using the 10X
Genomics platform.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Retrieve the dataset of interest</span></span>
<span><span class="fu">SeuratData</span><span class="fu">::</span><span class="fu">InstallData</span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the count matrix</span></span>
<span><span class="va">pbmc3k_counts</span> <span class="op">&lt;-</span> <span class="va">pbmc3k</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">counts</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="using-alignment-output">Using Alignment Output<a class="anchor" aria-label="anchor" href="#using-alignment-output"></a>
</h4>
<p>Alignment output comes in the form of three files containing the row
names (genes), column names (cell identifiers), and input values (gene
expression values) of the count matrix. These can be compiled using
ReadMtx, a function offered by Seurat which simplifies the count matrix
compilation into one function. This involves decompressing zipped files,
mapping feature names to HGNC gene symbols, and converting the matrix to
sparse format. ReadMtx output can be used directly for
DamageDetective.</p>
<p>We will use publicly available alignment data from the 10X Genomics
website for demonstration, specifically the ‘1k PBMCs from a healthy
donor (v3)’ data available here following the link named “Feature / cell
matrix (filtered)”.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the file paths relative to location on your device</span></span>
<span><span class="va">matrix_file</span> <span class="op">&lt;-</span> <span class="st">"~/Projects/demonstrations/matrix.mtx.gz"</span>     </span>
<span><span class="va">barcodes_file</span> <span class="op">&lt;-</span> <span class="st">"~/Projects/demonstrations/barcodes.tsv.gz"</span> </span>
<span><span class="va">features_file</span> <span class="op">&lt;-</span> <span class="st">"~/Projects/demonstrations/features.tsv.gz"</span> </span>
<span></span>
<span><span class="co"># Construct the sparse matrix</span></span>
<span><span class="va">alignment_counts</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ReadMtx.html" class="external-link">ReadMtx</a></span><span class="op">(</span></span>
<span>  mtx <span class="op">=</span> <span class="va">matrix_file</span>,</span>
<span>  cells <span class="op">=</span> <span class="va">barcodes_file</span>,</span>
<span>  features <span class="op">=</span> <span class="va">features_file</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="using-an-sce-object">Using an <code>sce</code> Object<a class="anchor" aria-label="anchor" href="#using-an-sce-object"></a>
</h4>
<p>The count matrix can be extracted from the assay slot of a Seurat
object. By default, this matrix is already in compressed column-oriented
sparse form and can be used as input for DamageDetective.</p>
<p>We will use the SeuratData package to retrieve a publicly available
Seurat object to demonstrate. This object stores the data from
peripheral blood mononuclear cells (PBMCs) sequenced using the 10X
Genomics platform.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Retrieve multisample dataset</span></span>
<span><span class="va">pbmc_sce</span> <span class="op">&lt;-</span> <span class="fu">scRNAseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/fetchDataset.html" class="external-link">fetchDataset</a></span><span class="op">(</span><span class="st">"kotliarov-pbmc-2020"</span>, <span class="st">"2024-04-18"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract sample of interest</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">pbmc_sce</span><span class="op">)</span></span>
<span><span class="va">sample_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="va">sample</span> <span class="op">==</span> <span class="st">"234_d0"</span><span class="op">)</span></span>
<span><span class="va">sample_sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sample_sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset and convert to sparse format</span></span>
<span><span class="va">pbmc_counts</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">pbmc_sce</span>, <span class="st">"counts"</span><span class="op">)</span></span>
<span><span class="va">sample_counts</span>  <span class="op">&lt;-</span> <span class="va">pbmc_counts</span><span class="op">[</span>, <span class="va">sample_sce</span><span class="op">]</span></span>
<span><span class="va">sample_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">sample_counts</span><span class="op">)</span></span>
<span><span class="va">sample_counts</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">sample_counts</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="parameter-selection">Parameter Selection<a class="anchor" aria-label="anchor" href="#parameter-selection"></a>
</h2>
<p>The <code>detect_damage</code> function requires only the count
matrix to run but accepts additional parameters. These parameters can be
divided into two categories:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Computational Parameters</strong>: Alter the computation and
result in different outputs.</li>
<li>
<strong>Aesthetic Parameters</strong>: Adjust how the user receives
the output but do not change the computation.</li>
</ol>
<div class="section level3">
<h3 id="computational-parameters">Computational Parameters<a class="anchor" aria-label="anchor" href="#computational-parameters"></a>
</h3>
<div class="section level4">
<h4 id="dataset-defined-parameters">Dataset-Defined Parameters<a class="anchor" aria-label="anchor" href="#dataset-defined-parameters"></a>
</h4>
<ul>
<li>
<strong><code>organism</code></strong>: Specifies the organism of
the data (default: <code>"Hsap"</code>).</li>
<li>
<strong><code>annotated_celltypes</code></strong>: Indicates whether
cell types are known (default: <code>FALSE</code>).</li>
<li>
<strong><code>ribosome_penalty</code></strong>: Accounts for the
unexpectedly low probability of ribosomal RNA loss in real single-cell
data.</li>
</ul>
<div class="section level5">
<h5 id="exploring-the-ribosome_penalty-parameter">Exploring the <code>ribosome_penalty</code> parameter<a class="anchor" aria-label="anchor" href="#exploring-the-ribosome_penalty-parameter"></a>
</h5>
<p>This parameter accounts for the unexpectedly low probability of
ribosomal RNA loss observed in real single cell data that must be
accounted for during the simulation of artificial cells.</p>
<p>The impact of changing <code>ribosome_penalty</code> can be explored
in the plots below. The idea is to see whether the artificial cells
describe the true cells, in other words, whether the coloured dots
superimpose the grey dots.</p>
<ul>
<li>Alignment output</li>
</ul>
<p>You will see that as you increase the penalty, i.e. go from values
closer to 1 to values closer to
0<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mo>*</mo></msup><annotation encoding="application/x-tex">^*</annotation></semantics></math>,
the coloured dots shift from an extreme position on the left-hand side
of the plot to a more central position. At what point along this range
true cells exist is dataset-dependent but generally lies closer to
0.</p>
<blockquote>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi></mi><mo>*</mo></msup><annotation encoding="application/x-tex">^*</annotation></semantics></math>
ribosome_penalty is a multiplicative reduction factor meaning a value of
1 is the same as introducing no penalty while values increasingly closer
to zero introduce increasingly greater reductions.</p>
</blockquote>
<div class="float">
<img src="../reference/figures/ribosome_penalty_alignment.svg" alt="A scatter plot showing the effect of ribosomal penalty on alignment output."><div class="figcaption">A scatter plot showing the effect of ribosomal
penalty on alignment output.</div>
</div>
<p><br></p>
<p>Plots like that shown above can be generated as follows where,</p>
<ul>
<li><p><code>count_matrix</code> requires your data in sparse matrix
form</p></li>
<li><p><code>ribosome_penalty</code> is a numeric between 0 and 1
specifying the ribosomal penalty</p></li>
<li><p><code>damage_proportion</code> is a number between 0 and 1
specifying the amount of artificial cells to create relative to the
input data (<em>setting this to a lower value makes the computation
faster</em>)</p></li>
<li><p><code>target_damage</code> is a vector specifying the lower and
upper levels of damage between which the artificial cells are generated.
Shifting this to higher values shifts the damaged cell profiles to more
intense levels that can be seen by a shift in colour, more red, as well
as a shift in position, higher up on the y axis.</p></li>
<li><p><code>plot_ribosomal_penalty</code> allows the plot to focus on
the ribosomal proportion rather than give a general overview of the QC
distributions.</p></li>
<li><p><code>seed</code> ensures the output can be reproduced.</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">penalty_plot</span> <span class="op">&lt;-</span> <span class="fu">DamageDetective</span><span class="fu">::</span><span class="fu"><a href="../reference/simulate_counts.html">simulate_counts</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">alignment_counts</span>,</span>
<span>  ribosome_penalty <span class="op">=</span> <span class="fl">0.01</span>,  </span>
<span>  damage_proportion <span class="op">=</span> <span class="fl">0.05</span>,  </span>
<span>  target_damage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>  plot_ribosomal_penalty <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="detection-vignette_files/figure-html/view_penalty-1.png" alt="Scatter plot showing the distribution of quality control  metrics in altered and unaltered counts. " width="700"></p>
<p><br></p>
</div>
</div>
<div class="section level4">
<h4 id="computing-the-ideal-ribosome_penalty">Computing the Ideal <code>ribosome_penalty</code><a class="anchor" aria-label="anchor" href="#computing-the-ideal-ribosome_penalty"></a>
</h4>
<p>From the above plots you can see how selecting an unideal
`ribosome_penalty` can generate damaged profiles that do not describe
the data well and, as a result, generate estimations of damage that do
not describe the data well.</p>
<p>Selecting a `ribosome_penalty` that simulates RNA loss in a way that
is relevant to the input data can be done as shown above, through trial
and error, or in an automated fashion using the `select_penalty`
function,</p>
<ul>
<li><p><code>penalty_range</code> setting a specific range of
<code>ribosome_penalty</code> estimates</p></li>
<li><p><code>max_penalty_trials</code> set a maximum number of attempts
allowed</p></li>
<li><p><code>penalty_step</code> setting the difference between each
estimate</p></li>
<li><p><code>return_output</code> returns a table showing the mean
distances for each estimate giving a better idea of how the estimates
differ.</p></li>
</ul>
<p>It also inherits parameters of the <code>simulate_counts</code>
function above including <code>damage_proportion</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_penalty</span> <span class="op">&lt;-</span> <span class="fu">DamageDetective</span><span class="fu">::</span><span class="fu"><a href="../reference/select_penalty.html">select_penalty</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">alignment_counts</span>,</span>
<span>  max_penalty_trials <span class="op">=</span> <span class="fl">3</span>, <span class="co"># shortened for vignette, default is 10</span></span>
<span>  seed <span class="op">=</span> <span class="fl">7</span>, </span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Testing penalty of 1e-05...</span></span>
<span><span class="co">#&gt; Testing penalty of 0.00501...</span></span>
<span><span class="co">#&gt; Testing penalty of 0.01001...</span></span>
<span><span class="co">#&gt; Maximum penalty trials reached (3). Stopping.</span></span>
<span><span class="va">selected_penalty</span></span>
<span><span class="co">#&gt; [1] 1e-05</span></span></code></pre></div>
<p>The output above reveals that the <code>select_penalty</code>
function runs through trial and error. This involves generating
artificial cells using different <code>ribosome_penalty</code> estimates
and evaluating the similarity of the resulting artificial cells to true
cells.</p>
<p>Here, ‘similarity’ refers to the shortest distance from each
artificial cell to a true cell in principal component (PC) space. The
mean shortest distances are taken for each level of damage simulated.
The penalty that generates the smallest mean shortest distance across
damage levels is selected as the ideal <code>ribosome_penalty</code>. In
other words, a penalty that generates artificial cells that are the most
similar to true cells over a wide range of damage levels is ideal.</p>
<p>Though not strictly necessary, the <code>select_penalty</code>
function can be adjusted to give a user more control over the testing
process. This includes,</p>
<ul>
<li><p><code>penalty_range</code> setting a specific range of
<code>ribosome_penalty</code> estimates</p></li>
<li><p><code>max_penalty_trials</code> set a maximum number of attempts
allowed</p></li>
<li><p><code>penalty_step</code> setting the difference between each
estimate</p></li>
<li><p><code>return_output</code> returns a table showing the mean
distances for each estimate giving a better idea of how the estimates
differ.</p></li>
</ul>
<p>It also inherits parameters of the <code>simulate_counts</code>
function above including <code>damage_proportion</code>.</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="undefined-computational-parameters">Undefined Computational Parameters<a class="anchor" aria-label="anchor" href="#undefined-computational-parameters"></a>
</h4>
<p>The remainder of the computational parameters involve selections that
cannot be determined directly from the dataset and are up to user
discretion. To support this, it may be helpful for a user to understand
how the <code>detect_damage</code> function works.</p>
<p>Ultimately, <code>detect_damage</code> outputs a score from 0 to 1
indicating the estimated damage level in each cell. This is found by
comparing the cells to simulated cells where the extent of damage is
known.</p>
<p>Damage is simulated by the loss of cytoplasmic RNA, a consequence
that follows from a loss in plasma membrane integrity-a defining
principle of damage used in many cell viability assays.
<code>DamageDetective</code> assumes that the proportion of RNA lost is
directly related to the extent of cellular damage.</p>
<p>In a complementary workflow to that of <code>select_penalty</code>,
<code>detect_damage</code> estimates the damage level of each true cell
by measuring its similarity to artificially damaged cells. Using the PC
embeddings, a set of cells with the highest similarity to each true
cell, the nearest neighbours (NN), is collected. The proportions of
nearest neighbours that originate from damage level are found. The
damage level of the artificial set to which the true cell shows the
highest proportion of nearest neighbours is selected as the cell’s
estimated damage level. This is scaled to score the cells in each set in
a way that reflects the relative differences in proportion.</p>
<p>Now, looking at the parameters:</p>
<ul>
<li><strong><code>filter_threshold</code></strong></li>
</ul>
<p>A value between 0 and 1 to determine the level of damage, or
proportion of RNA loss, above which cells will be excluded. By default,
<code>DamageDetective</code> offers the threshold of <code>0.7</code>.
Values greater than <code>0.7</code> reflect more permissive filtering
while those closer to <code>0</code> reflect more stringent filtering.
We recommend the default for all cases but suggest that if adjustments
are made, they are informed by inspecting the output
<code>detect_damage</code> plots, <code>generate_plot = TRUE</code>.</p>
<ul>
<li><strong><code>damage_levels</code></strong></li>
</ul>
<p>This describes the number of distinct sets of artificial cells
simulated, each with a defined level of damage. This parameter requires
a balance between generating enough damage sets to give informative
estimates of damage i.e. not 0 to 1, and not putting the computation
under strain by creating redundant sets.</p>
<p>Of the default options, we find 5 sets strikes an ideal balance.
Going above 5 becomes computationally challenging but could be done for
small datasets. Going below 5, while not ideal, may be necessary for
large datasets (over 10 000 cells).</p>
<p>While shifting from the default options is not recommended, it is
possible should a user feel it necessary through specification in a
list,</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">damage_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  pANN_50 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>  pANN_100 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<ul>
<li><strong><code>kN</code></strong></li>
</ul>
<p>This describes the size of the set of nearest neighbours that is
found for each true cell. This value defaults to a fifth of the size of
the input dataset. This parameter is not as essential as others since
changes made here introduce change to all artificial cells in a relative
way.</p>
<p>It should be noted that increasing the size of the set too greatly
increases the computational requirements of the function and have not
been found to dircetly improve output. For this reason, anything greater
than a third of the size of the input data is discouraged.</p>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="aesthetic-parameters">Aesthetic Parameters<a class="anchor" aria-label="anchor" href="#aesthetic-parameters"></a>
</h3>
<ul>
<li>
<strong><code>filter_counts</code></strong>: Determines the output
format (filtered matrix or dataframe).</li>
<li>
<strong><code>palette</code></strong>: Sets the color palette for
damaged cells in plots.</li>
</ul>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="damaged-cell-detection">Damaged Cell Detection<a class="anchor" aria-label="anchor" href="#damaged-cell-detection"></a>
</h2>
<p>Once the data is in sparse matrix form and an appropriate
<code>ribosome_penalty</code> is selected, the
<code>detect_damage</code> function can be run.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run detection</span></span>
<span><span class="va">detection_output</span> <span class="op">&lt;-</span> <span class="fu">DamageDetective</span><span class="fu">::</span><span class="fu"><a href="../reference/detect_damage.html">detect_damage</a></span><span class="op">(</span></span>
<span>  count_matrix <span class="op">=</span> <span class="va">alignment_counts</span>, </span>
<span>  ribosome_penalty <span class="op">=</span> <span class="va">selected_penalty</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Simulating 1e-05 and 0.08 RNA loss...</span></span>
<span><span class="co">#&gt; Simulating 0.1 and 0.3 RNA loss...</span></span>
<span><span class="co">#&gt; Simulating 0.3 and 0.5 RNA loss...</span></span>
<span><span class="co">#&gt; Simulating 0.5 and 0.7 RNA loss...</span></span>
<span><span class="co">#&gt; Simulating 0.7 and 0.9 RNA loss...</span></span>
<span><span class="co">#&gt; Computing pANN...</span></span></code></pre></div>
<p><img src="detection-vignette_files/figure-html/run_detection-1.png" alt="Scatter plot showing the outcome damage detection where cells of the input data are coloured according to the estimated level of damage with grey being no damage,  purple being light damage, and red being heavy damage." width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># View output</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">detection_output</span><span class="op">$</span><span class="va">output</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                    DamageDetective DamageDetective_filter</span></span>
<span><span class="co">#&gt; AAACCCAAGGAGAGTA-1     0.063257907                   cell</span></span>
<span><span class="co">#&gt; AAACGCTTCAGCCCAG-1     0.024193023                   cell</span></span>
<span><span class="co">#&gt; AAAGAACAGACGACTG-1     0.059537442                   cell</span></span>
<span><span class="co">#&gt; AAAGAACCAATGGCAG-1     0.181250000                   cell</span></span>
<span><span class="co">#&gt; AAAGAACGTCTGCAAT-1     0.007450930                   cell</span></span>
<span><span class="co">#&gt; AAAGGATAGTAGACAT-1     0.005590698                   cell</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">detection_output</span><span class="op">$</span><span class="va">output</span><span class="op">$</span><span class="va">DamageDetective_filter</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    cell damaged </span></span>
<span><span class="co">#&gt;    1150      72</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] patchwork_1.3.0       ggplot2_3.5.1         Seurat_5.2.1         </span></span>
<span><span class="co">#&gt; [4] SeuratObject_5.0.2    sp_2.2-0              DamageDetective_1.0.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     jsonlite_2.0.0         magrittr_2.0.3        </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.1-3   farver_2.1.2           rmarkdown_2.29        </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.5               ragg_1.3.3             vctrs_0.6.5           </span></span>
<span><span class="co">#&gt;  [10] ROCR_1.0-11            spatstat.explore_3.4-2 rstatix_0.7.2         </span></span>
<span><span class="co">#&gt;  [13] htmltools_0.5.8.1      broom_1.0.8            Formula_1.2-5         </span></span>
<span><span class="co">#&gt;  [16] sass_0.4.9             sctransform_0.4.1      parallelly_1.43.0     </span></span>
<span><span class="co">#&gt;  [19] KernSmooth_2.23-26     bslib_0.9.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">#&gt;  [22] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">#&gt;  [25] plotly_4.10.4          zoo_1.8-13             cachem_1.1.0          </span></span>
<span><span class="co">#&gt;  [28] igraph_2.1.4           mime_0.13              lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt;  [31] pkgconfig_2.0.3        Matrix_1.7-2           R6_2.6.1              </span></span>
<span><span class="co">#&gt;  [34] fastmap_1.2.0          fitdistrplus_1.2-2     future_1.34.0         </span></span>
<span><span class="co">#&gt;  [37] shiny_1.10.0           digest_0.6.37          colorspace_2.1-1      </span></span>
<span><span class="co">#&gt;  [40] tensor_1.5             RSpectra_0.16-2        irlba_2.3.5.1         </span></span>
<span><span class="co">#&gt;  [43] textshaping_1.0.0      ggpubr_0.6.0           labeling_0.4.3        </span></span>
<span><span class="co">#&gt;  [46] progressr_0.15.1       spatstat.sparse_3.1-0  httr_1.4.7            </span></span>
<span><span class="co">#&gt;  [49] polyclip_1.10-7        abind_1.4-8            compiler_4.4.3        </span></span>
<span><span class="co">#&gt;  [52] withr_3.0.2            backports_1.5.0        carData_3.0-5         </span></span>
<span><span class="co">#&gt;  [55] fastDummies_1.7.5      ggsignif_0.6.4         MASS_7.3-64           </span></span>
<span><span class="co">#&gt;  [58] tools_4.4.3            lmtest_0.9-40          httpuv_1.6.15         </span></span>
<span><span class="co">#&gt;  [61] future.apply_1.11.3    goftest_1.2-3          glue_1.8.0            </span></span>
<span><span class="co">#&gt;  [64] nlme_3.1-167           promises_1.3.2         grid_4.4.3            </span></span>
<span><span class="co">#&gt;  [67] Rtsne_0.17             cluster_2.1.8          reshape2_1.4.4        </span></span>
<span><span class="co">#&gt;  [70] generics_0.1.3         gtable_0.3.6           spatstat.data_3.1-6   </span></span>
<span><span class="co">#&gt;  [73] tidyr_1.3.1            data.table_1.17.0      car_3.1-3             </span></span>
<span><span class="co">#&gt;  [76] spatstat.geom_3.3-6    RcppAnnoy_0.0.22       ggrepel_0.9.6         </span></span>
<span><span class="co">#&gt;  [79] RANN_2.6.2             pillar_1.10.1          stringr_1.5.1         </span></span>
<span><span class="co">#&gt;  [82] spam_2.11-1            RcppHNSW_0.6.0         later_1.4.1           </span></span>
<span><span class="co">#&gt;  [85] splines_4.4.3          dplyr_1.1.4            lattice_0.22-6        </span></span>
<span><span class="co">#&gt;  [88] survival_3.8-3         deldir_2.0-4           tidyselect_1.2.1      </span></span>
<span><span class="co">#&gt;  [91] miniUI_0.1.1.1         pbapply_1.7-2          knitr_1.50            </span></span>
<span><span class="co">#&gt;  [94] gridExtra_2.3          scattermore_1.2        xfun_0.52             </span></span>
<span><span class="co">#&gt;  [97] matrixStats_1.5.0      stringi_1.8.7          lazyeval_0.2.2        </span></span>
<span><span class="co">#&gt; [100] yaml_2.3.10            evaluate_1.0.3         codetools_0.2-20      </span></span>
<span><span class="co">#&gt; [103] tibble_3.2.1           cli_3.6.4              uwot_0.2.3            </span></span>
<span><span class="co">#&gt; [106] xtable_1.8-4           reticulate_1.42.0      systemfonts_1.2.1     </span></span>
<span><span class="co">#&gt; [109] munsell_0.5.1          jquerylib_0.1.4        Rcpp_1.0.14           </span></span>
<span><span class="co">#&gt; [112] globals_0.16.3         spatstat.random_3.3-3  png_0.1-8             </span></span>
<span><span class="co">#&gt; [115] spatstat.univar_3.1-2  parallel_4.4.3         pkgdown_2.1.1         </span></span>
<span><span class="co">#&gt; [118] dotCall64_1.2          listenv_0.9.1          viridisLite_0.4.2     </span></span>
<span><span class="co">#&gt; [121] scales_1.3.0           ggridges_0.5.6         purrr_1.0.4           </span></span>
<span><span class="co">#&gt; [124] rlang_1.1.5            cowplot_1.1.3</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Amezquita R, et al. (2020). “Orchestrating single-cell analysis with
Bioconductor.” <em>Nature Methods</em>, <em>17</em>, 137-145.</li>
<li>Hao et al. (2023). <em>Seurat V5</em>. Nature Biotechnology.</li>
<li>Satija R, et al. (2025). <em>SeuratData: Install and Manage Seurat
Datasets</em>. R package version 0.2.2.9002.</li>
<li>Pedersen T (2024). <em>patchwork: The Composer of Plots</em>. R
package version 1.3.0.</li>
<li>Wickham. <em>ggplot2: Elegant Graphics for Data Analysis</em>.
Springer-Verlag New York, 2016.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alicen Henning.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
